### PR TITLE
Fix PKCS#7 inner contents verification

### PR DESCRIPTION
**Pull request by olszomal**
The general syntax for **content** according to [RFC 2315](https://datatracker.ietf.org/doc/html/rfc2315#section-7) associates a content type with content.
```
ContentInfo ::= SEQUENCE {
     contentType ContentType,
     content
       [0] EXPLICIT ANY DEFINED BY contentType OPTIONAL }

   ContentType ::= OBJECT IDENTIFIER
 ```  
The data content type is just an octet string.
```
Data ::= OCTET STRING
```

[Microsoft SignTool](https://learn.microsoft.com/en-us/windows/win32/seccrypto/signtool) or [osslsigncode ](https://github.com/mtrojnar/osslsigncode) tool can digitally sign files.
Here is an example of parsing PKCS#7 signature.
`p7->d.sign->contents` starts at an offset 43 and has the `V_ASN1_SEQUENCE` type.
```
$ openssl asn1parse -i -inform pem -in sig.pem
    0:d=0  hl=4 l=1571 cons: SEQUENCE
    4:d=1  hl=2 l=   9 prim:  OBJECT            :pkcs7-signedData
   15:d=1  hl=4 l=1556 cons:  cont [ 0 ]
   19:d=2  hl=4 l=1552 cons:   SEQUENCE
   23:d=3  hl=2 l=   1 prim:    INTEGER           :01
   26:d=3  hl=2 l=  15 cons:    SET
   28:d=4  hl=2 l=  13 cons:     SEQUENCE
   30:d=5  hl=2 l=   9 prim:      OBJECT            :sha256
   41:d=5  hl=2 l=   0 prim:      NULL
   43:d=3  hl=2 l=  92 cons:    SEQUENCE
   45:d=4  hl=2 l=  10 prim:     OBJECT            :1.3.6.1.4.1.311.2.1.4
   57:d=4  hl=2 l=  78 cons:     cont [ 0 ]
   59:d=5  hl=2 l=  76 cons:      SEQUENCE
   61:d=6  hl=2 l=  23 cons:       SEQUENCE
   63:d=7  hl=2 l=  10 prim:        OBJECT            :1.3.6.1.4.1.311.2.1.15
   75:d=7  hl=2 l=   9 cons:        SEQUENCE
   77:d=8  hl=2 l=   1 prim:         BIT STRING
   80:d=8  hl=2 l=   4 cons:         cont [ 0 ]
   82:d=9  hl=2 l=   2 cons:          cont [ 2 ]
   84:d=10 hl=2 l=   0 prim:           cont [ 0 ]
   86:d=6  hl=2 l=  49 cons:       SEQUENCE
   88:d=7  hl=2 l=  13 cons:        SEQUENCE
   90:d=8  hl=2 l=   9 prim:         OBJECT            :sha256
  101:d=8  hl=2 l=   0 prim:         NULL
  103:d=7  hl=2 l=  32 prim:        OCTET STRING      [HEX DUMP]:2D2C7B382C8163A419B9FF214A7B651C33F9EA43335907F11377290C5158A7A4
  137:d=3  hl=4 l= 913 cons:    cont [ 0 ]
  141:d=4  hl=4 l= 909 cons:     SEQUENCE
  145:d=5  hl=4 l= 629 cons:      SEQUENCE
  149:d=6  hl=2 l=   3 cons:       cont [ 0 ]
  151:d=7  hl=2 l=   1 prim:        INTEGER           :02
  154:d=6  hl=2 l=  20 prim:       INTEGER           :1A2F945F4D1E9B68F8BFC6C9D6C707B03E35FDC7
  (...)
```  

In the [PKCS7_verify()](https://www.openssl.org/docs/manmaster/man3/PKCS7_verify.html) function, the BIO *indata parameter refers to the signed data if the content is detached from p7. Otherwise, indata should be NULL, and then the signed data must be in p7. The content is written to the BIO out unless it is NULL.

Currently, the only supported types for `p7->d.sign->contents` are `NID_pkcs7_data` or `V_ASN1_OCTET_STRING`, due to restrictions in the `PKCS7_get_octet_string()` function.

My PR addresses this by adding support for `p7->d.sign->contents` with the `V_ASN1_SEQUENCE` type.

- [x] tests are added or updated

CLA: regular

### CODE REVIEW

**Comment by t8m** [comment]
CI is relevant

**Comment by olszomal** [comment]
This PR should be triaged as a bug, not a feature, I think.

According to the `PKCS7_verify()` manual:
```
indata refers to the signed data if the content is detached from p7. Otherwise indata should be NULL, and then the signed data must be in p7.
```

Without this PR, the **inner content** must be passed as **detached content** with the _indata_ parameter.

This is a workaround to put the **inner content** into 'indata' parameter:

``` C
    BIO *bio = NULL;
    if (p7->d.sign->contents->d.other->type == V_ASN1_SEQUENCE) {
        /* only verify the contents of the sequence */
        unsigned char *data = p7->d.sign->contents->d.other->value.sequence->data;
        int len = p7->d.sign->contents->d.other->value.sequence->length;
        int hdrlen = ASN1_object_size(0, len, V_ASN1_SEQUENCE) - len;
        bio = BIO_new_mem_buf(data + hdrlen, length - hdrlen);
    }
    if (!PKCS7_verify(p7, NULL, store, bio, NULL, 0)) {
        /* Verification failure */
        (...)
    }
```

**Comment by olszomal** [comment]
Please let me know if you need any help from me.

**Comment by t8m** [comment]
The CI failures must be fixed.

File: crypto/pkcs7/pk7_doit.c
Comment (t8m) [inline crypto/pkcs7/pk7_doit.c:61]: Missing empty line after declarations.

File: crypto/pkcs7/pk7_doit.c
Comment (t8m) [inline crypto/pkcs7/pk7_doit.c:63]: pointers must be compared against NULL

File: crypto/pkcs7/pk7_doit.c
Comment (t8m) [inline crypto/pkcs7/pk7_doit.c:72]: Missing empty line after declarations

**Review by t8m — APPROVED** [review]

**Comment by mtrojnar** [comment]
Do I correctly understand that this PR is ready for merging?

**Comment by t8m** [comment]
No, this still needs a second review.

**Comment by beldmit** [comment]
Could the tests use stored file instead of hard-coded buffers?

**Review by mattcaswell — DISMISSED** [review]

**Comment by mtrojnar** [comment]
> Could the tests use stored file instead of hard-coded buffers?

I guess it could, but refactoring of an existing test for unrelated functionality is beyond the scope of this PR, and using different methods of retrieving the date in a single test source file would be a bad coding style.

File: crypto/pkcs7/pk7_doit.c
Comment (olszomal) [reply crypto/pkcs7/pk7_doit.c:66]: Flag `ASN1_STRING_FLAG_NDEF` is currently used by openssl-smime.
https://github.com/openssl/openssl/blob/bdba075c1962b7c69b6f40ebef9e0bf1f1325dd8/include/openssl/asn1.h.in#L139-L144

`ASN1_OCTET_STRING_dup(os)` can not be done if `ASN1_STRING_FLAG_NDEF` flag is set, because `os` does not contain any data:
os->flags = 0x0010
os->length = 0
os->data == NULL

File: crypto/pkcs7/pk7_doit.c
Comment (olszomal) [reply crypto/pkcs7/pk7_doit.c:405]: `ASN1_OCTET_STRING_free(os)` can not be done if `ASN1_STRING_FLAG_NDEF` flag is set

File: crypto/pkcs7/pk7_doit.c
Comment (olszomal) [reply crypto/pkcs7/pk7_doit.c]: I’ve changed it to the `ASN1_get_object` function.

**Review by t8m — APPROVED** [review]

File: crypto/pkcs7/pk7_doit.c
Comment (mattcaswell) [inline crypto/pkcs7/pk7_doit.c:80]: I'm still confused by this....are we saying that in the case of "type_is_other", the contained data must be a sequence within a sequence?? Why? How do we know that this will be generically true for all "other" types that we might want to have in the future?

Comment (olszomal) [reply crypto/pkcs7/pk7_doit.c:80]: The existing OpenSSL implementation supports only MIME content. This PR enables support for PKCS#7 content.

[CMS](https://datatracker.ietf.org/doc/html/rfc5652#section-5.2) defines eContent as:
```
    eContent [0] EXPLICIT OCTET STRING OPTIONAL
```

In S/MIME v2, S/MIME v3, and S/MIME v3.1, the MIME content is placed in an OCTET STRING.

[PKCS#7](https://datatracker.ietf.org/doc/html/rfc2315#section-7) defines content as:
```
    content [0] EXPLICIT ANY DEFINED BY contentType OPTIONAL
```

The contentType indicates the type of content.
```
    ContentType ::= OBJECT IDENTIFIER
```

The data content type is just an octet string. The interpretation is left to the application. It shall have ASN.1 type Data:
```
    Data ::= OCTET STRING
```

See [rfc5652#section-5.2.1](https://datatracker.ietf.org/doc/html/rfc5652#section-5.2.1)

Comment (t8m) [reply crypto/pkcs7/pk7_doit.c:80]: ping @mattcaswell

Comment (mtrojnar) [reply crypto/pkcs7/pk7_doit.c:80]: ping @mattcaswell

Comment (mattcaswell) [reply crypto/pkcs7/pk7_doit.c:80]: Sorry for my lack of response.

This answer does not seem to address my question.

IIUC, PKCS#7 only defines content types for data, signedData, envelopedData, signedAndEnvelopedData, digestedData, and encryptedData. Where "data" in particular is an OCTET STRING.

The function `PKCS7_get_octet_string` attempts to get that octet string. It checks that the content type is "data" and returns the OCTET STRING associated with it. If the content type is something other than one of the content types that we recognise then we don't really know what to do with it - but if the underlying contents happen to be an OCTET STRING anyway then `PKCS7_get_octet_string` can still do something sensible. This seems somewhat dubious behaviour IMO - but ok its existing legacy code.

What you seem to be attempting to do here is something different. We've got a content type that we don't recognise. You seem to be checking if it happens to contain a sequence. If it does then you look inside that sequence to see if we have another sequence. If it does then you convert that sequence data into an OCTET STRING and return it.

This seems significantly more dubious than the currently existing dubious behaviour. We've got a content type that we don't know what it is, checking to see if it has some arbitrary structure to the unknown data and then converting a bit of that unknown data from structured data into an OCTET STRING. This behaviour seems entirely arbitrary and non-sensical. Presumably it is done this way to get this one use case to work - but it doesn't seem to make sense to me from a generic perspective.

It seems what you are really trying to do is add support for a new content type. If we know what the content type is then we can know exactly what structure the content is expected to have and respond to it accordingly. Without knowing what the content type is we are just guessing what the right thing to do is.

However adding new content types to the PKCS7 implementation seems entirely wrong. This is all legacy stuff. It seems to me that the OpenSSL CMS implementation would be the right place to do this kind of thing.

Comment (vdukhovni) [reply crypto/pkcs7/pk7_doit.c:80]: > [@mattcaswell] What you seem to be attempting to do here is something different. We've got a content type that we don't recognise. You seem to be checking if it happens to contain a sequence. If it does then you look inside that sequence to see if we have another sequence. If it does then you convert that sequence data into an OCTET STRING and return it.

FWIW (without prejudice on whether the overall approach is or isn't sound), my reading of the code and intent is that there is only **one** layer of SEQUENCE here. That is, the data returned is the serialisation of the *topmost* SEQUENCE that makes up the PKCS#7 content of the custom type. This is at the same level as the octet-string, differing only in the tag (SEQUENCE vs. OCTET-STRING).

Whether we want to add new PKCS#7 functionality, whether there are new memory management concerns, ... are still review-worthy questions, but on this one point perhaps there's a simple misunderstanding?

Comment (vdukhovni) [reply crypto/pkcs7/pk7_doit.c:80]: I discussed this issue with @mattcaswell and @t8m and the crux of the issue the code in https://github.com/openssl/openssl/blob/d3fea0256038d05a76a8be9152d3cf2abf32c3cf/crypto/pkcs7/pk7_doit.c#L75-L85

Matt's reading of those lines was that you're looking for a second *nested* sequence inside the `content` of the PKCS#7 `ContentInfo`, when the `content` is some sort of sequence. But, on reading these lines more closely, I'm convinced that all that's happening is that the DER encoding of that same outer sequence, stored at `p7->d.other->value.sequence->data` is decoded to only to discard the tag and length, with a sanity check that we got the *expected* tag, and it is then the (presumably) DER payload of the sequence that is returned in `os`.

If that reading is correct, I think @mattcaswell made an honest mistake, misreading the code. Given what PKCS#7 digests are expected to cover (`content` sans tag and length) this part of the PR looks reasonable to me.

Comment (mattcaswell) [reply crypto/pkcs7/pk7_doit.c:80]: > Matt's reading of those lines was that you're looking for a second nested sequence inside the content of the PKCS#7 ContentInfo, when the content is some sort of sequence. But, on reading these lines more closely, I'm convinced that all that's happening is that the DER encoding of that same outer sequence, stored at p7->d.other->value.sequence->data is decoded to only to discard the tag and length, with a sanity check that we got the expected tag, and it is then the (presumably) DER payload of the sequence that is returned in os.

You are right!! Wow this is really non-obvious and not what I expected. It took firing up a debugger and examining data structures to really figure this out. Perhaps a comment here to explain what is really going on would be a good idea?

With that new understanding this code makes a lot more sense to me. The `ASN1_get_object` call then does return the payload length of the sequence. But that seems consistent with the PKCS7 spec:

> Only the contents octets of the DER encoding of that field are digested, not the identifier octets or the length octets.

So I'm ready to "resolve" this comment - except for the residual suggestion to add a comment here somewhere.

File: crypto/pkcs7/pk7_doit.c
Comment (mattcaswell) [reply crypto/pkcs7/pk7_doit.c:66]: Hmm...does this actually fail? Or does it just end up with a non-NULL `os->data`? I've not tried it but my reading of the code suggests this might the case. I wonder if this is some deficiency in `ASN1_OCTET_STRING_dup()`? Should it really be able to handle this properly? One workaround might be to call `ASN1_OCTET_STRING_dup()` unconditionally, and then call `ASN1_STRING_set0((ASN1_STRING *)os, NULL, 0)` on the output in the `ASN1_STRING_FLAG_NDEF` case.

It seems preferable to have this function always return an allocated `ASN1_OCTET_STRING` that the caller is responsible for freeing - rather than have some obscure cases when this does not happen (especially since you named the function with `get1` which strongly implies the "caller must free" semantics)

Comment (mtrojnar) [reply crypto/pkcs7/pk7_doit.c]: @olszomal has implemented calling `ASN1_OCTET_STRING_dup()` unconditionally, and it indeed simplifies calling `ASN1_OCTET_STRING_free()`.

File: crypto/pkcs7/pk7_doit.c
Comment (mtrojnar) [reply crypto/pkcs7/pk7_doit.c]: @olszomal addressed this comment in the recent force-push.

File: crypto/pkcs7/pk7_doit.c
Comment (mtrojnar) [reply crypto/pkcs7/pk7_doit.c]: > So I'm ready to "resolve" this comment - except for the residual suggestion to add a comment here somewhere.

@olszomal has added some comments to clarify the purpose of non-trivial changes.

**Review by t8m — APPROVED** [review]

**Comment by t8m** [comment]
ping @openssl/committers for the second review

**Comment by olszomal** [comment]
Thanks for the review! I've applied the suggested changes, and the code works as expected. Let me know if anything else is needed.

**Review by t8m — CHANGES_REQUESTED** [review]

**Review by t8m — DISMISSED** [review]

**Review by hlandau — CHANGES_REQUESTED** [review]
This warrants a comment in CHANGES.md, otherwise LGTM

File: crypto/pkcs7/pk7_doit.c
Comment (vdukhovni) [inline crypto/pkcs7/pk7_doit.c:397]:
```suggestion
            if (bio != NULL) {
                BIO_set_mem_eof_return(bio, 0);
                if (BIO_write(bio, os->data, os->length) != os->length) {
                    BIO_free_all(bio);
                    bio = NULL;
                }
            }
```
Memory BIOs default to signalling a retriable error at the end of the available data, rather than return EOF. I don't think that's what we want here. The caller should see a hard EOF after reading the entirety `os->data`.
See similar code in the `else` branch.

**Review by vdukhovni — CHANGES_REQUESTED** [review]

File: crypto/pkcs7/pk7_doit.c
Comment (olszomal) [reply crypto/pkcs7/pk7_doit.c]: Good point, it's been fixed.

**Comment by t8m** [comment]
@olszomal Could you please add a CHANGES.md entry?

**Review by vdukhovni — APPROVED** [review]

**Review by t8m — APPROVED** [review]

**Comment by t8m** [comment]
@vdukhovni could you please re-approve

**Comment by vdukhovni** [comment]
By the way, do you know whether Emilia's comment is still pertinent: https://github.com/openssl/openssl/blob/f1607c8a2c04bcb95ddb2e6fc4e0aaec9729929b/crypto/pkcs7/pk7_doit.c#L863-L870

File: crypto/pkcs7/pk7_doit.c
Comment (vdukhovni) [inline crypto/pkcs7/pk7_doit.c:68]: Can you comment on why clearing the data is necessary?  Is it just to avoid looking at the NDEF flag later, because empty outputs are handled just like NDEF outputs? Or is there something else non-obvious going on?

**Review by vdukhovni — APPROVED** [review]

File: crypto/pkcs7/pk7_doit.c
Comment (olszomal) [reply crypto/pkcs7/pk7_doit.c:68]: It simplifies calling ASN1_OCTET_STRING_free(), see https://github.com/openssl/openssl/pull/22575#discussion_r1675602857

**Comment by olszomal** [comment]
> By the way, do you know whether Emilia's comment is still pertinent:
> 
> https://github.com/openssl/openssl/blob/f1607c8a2c04bcb95ddb2e6fc4e0aaec9729929b/crypto/pkcs7/pk7_doit.c#L863-L870

No, Emilia's comment is no longer pertinent. I have verified that the path where os != NULL is now executed in cases involving PKCS#7 inner contents.

**Comment by vdukhovni** [comment]
> > By the way, do you know whether Emilia's comment is still pertinent:
> > https://github.com/openssl/openssl/blob/f1607c8a2c04bcb95ddb2e6fc4e0aaec9729929b/crypto/pkcs7/pk7_doit.c#L863-L870
> 
> No, Emilia's comment is no longer pertinent. I have verified that the path where os != NULL is now executed in cases involving PKCS#7 inner contents.

I think Emilia was trying to say that the test `os == NULL` would never fire, were it not for some bug, so the question is not whether `os != NULL`, but whether it can be NULL here in some plausible error situation...

File: crypto/pkcs7/pk7_doit.c
Comment (olszomal) [reply crypto/pkcs7/pk7_doit.c:68]: Removing the line `ASN1_STRING_set0(osdup, NULL, 0);` causes the following memory leak when running the` openssl smime -sign` command:
```
==84352== HEAP SUMMARY:
==84352==     in use at exit: 1 bytes in 1 blocks
==84352==   total heap usage: 9,915 allocs, 9,914 frees, 775,183 bytes allocated
==84352== 
==84352== 1 bytes in 1 blocks are definitely lost in loss record 1 of 1
==84352==    at 0x484280F: malloc (vg_replace_malloc.c:442)
==84352==    by 0x4C0D105: CRYPTO_malloc (mem.c:202)
==84352==    by 0x4C0D270: CRYPTO_realloc (mem.c:295)
==84352==    by 0x4A91377: ASN1_STRING_set (asn1_lib.c:314)
==84352==    by 0x4A911EB: ASN1_STRING_copy (asn1_lib.c:262)
==84352==    by 0x4A91276: ASN1_STRING_dup (asn1_lib.c:279)
==84352==    by 0x4A889D3: ASN1_OCTET_STRING_dup (a_octet.c:16)
==84352==    by 0x4CFC96F: pkcs7_get1_data (pk7_doit.c:64)
==84352==    by 0x4CFD1CC: PKCS7_dataInit (pk7_doit.c:280)
==84352==    by 0x4CFB950: pk7_cb (pk7_asn1.c:46)
==84352==    by 0x4A97103: BIO_new_NDEF (bio_ndef.c:99)
==84352==    by 0x4A92D3F: i2d_ASN1_bio_stream (asn_mime.c:77)
==84352== 
==84352== LEAK SUMMARY:
==84352==    definitely lost: 1 bytes in 1 blocks
==84352==    indirectly lost: 0 bytes in 0 blocks
==84352==      possibly lost: 0 bytes in 0 blocks
==84352==    still reachable: 0 bytes in 0 blocks
==84352==         suppressed: 0 bytes in 0 blocks
==84352== 
==84352== ERROR SUMMARY: 1 errors from 1 contexts (suppressed: 0 from 0)

```

**Comment by olszomal** [comment]
> > No, Emilia's comment is no longer pertinent. I have verified that the path where os != NULL is now executed in cases involving PKCS#7 inner contents.
> 
> I think Emilia was trying to say that the test `os == NULL` would never fire, were it not for some bug, so the question is not whether `os != NULL`, but whether it can be NULL here in some plausible error situation...

So, in my opinion, this issue seems to be outside the scope of this PR

File: crypto/pkcs7/pk7_doit.c
Comment (vdukhovni) [reply crypto/pkcs7/pk7_doit.c:68]: > Removing the line `ASN1_STRING_set0(osdup, NULL, 0);` causes the following memory leak when running the` openssl smime -sign` command:
> 
> ```
> ==84352== HEAP SUMMARY:
> ==84352==     in use at exit: 1 bytes in 1 blocks
> ==84352==   total heap usage: 9,915 allocs, 9,914 frees, 775,183 bytes allocated
> ==84352== 
> ==84352== 1 bytes in 1 blocks are definitely lost in loss record 1 of 1
> ==84352==    at 0x484280F: malloc (vg_replace_malloc.c:442)
> ==84352==    by 0x4C0D105: CRYPTO_malloc (mem.c:202)
> ==84352==    by 0x4C0D270: CRYPTO_realloc (mem.c:295)
> ==84352==    by 0x4A91377: ASN1_STRING_set (asn1_lib.c:314)
> ==84352==    by 0x4A911EB: ASN1_STRING_copy (asn1_lib.c:262)
> ==84352==    by 0x4A91276: ASN1_STRING_dup (asn1_lib.c:279)
> ==84352==    by 0x4A889D3: ASN1_OCTET_STRING_dup (a_octet.c:16)
> ==84352==    by 0x4CFC96F: pkcs7_get1_data (pk7_doit.c:64)
> ==84352==    by 0x4CFD1CC: PKCS7_dataInit (pk7_doit.c:280)
> ==84352==    by 0x4CFB950: pk7_cb (pk7_asn1.c:46)
> ==84352==    by 0x4A97103: BIO_new_NDEF (bio_ndef.c:99)
> ==84352==    by 0x4A92D3F: i2d_ASN1_bio_stream (asn_mime.c:77)
> ==84352== 
> ==84352== LEAK SUMMARY:
> ==84352==    definitely lost: 1 bytes in 1 blocks
> ==84352==    indirectly lost: 0 bytes in 0 blocks
> ==84352==      possibly lost: 0 bytes in 0 blocks
> ==84352==    still reachable: 0 bytes in 0 blocks
> ==84352==         suppressed: 0 bytes in 0 blocks
> ==84352== 
> ==84352== ERROR SUMMARY: 1 errors from 1 contexts (suppressed: 0 from 0)
> ```

Interesting! This could be argued to be a bug in `ASN1_STRING_copy`, since it copies the flags (including NDEF), and should know to not copy any data, in that case, since it won't be freed?  Or did I misunderstand something here?

**Comment by vdukhovni** [comment]
> > > No, Emilia's comment is no longer pertinent. I have verified that the path where os != NULL is now executed in cases involving PKCS#7 inner contents.
> > 
> > 
> > I think Emilia was trying to say that the test `os == NULL` would never fire, were it not for some bug, so the question is not whether `os != NULL`, but whether it can be NULL here in some plausible error situation...
> 
> So, in my opinion, this issue seems to be outside the scope of this PR

Yes, quite likely!  I was just wondering whether in working on the PR you have new insight on this otherwise terse comment, and perhaps happened to make it obsolete.  I am guessing not then...

File: crypto/pkcs7/pk7_doit.c
Comment (vdukhovni) [reply crypto/pkcs7/pk7_doit.c:68]: @olszomal What do you think? Is the need to explicitly zero out the duplicated string to avoid a leak a bug in how NDEF strings are duplicated?

Comment (t8m) [reply crypto/pkcs7/pk7_doit.c:68]: I'd say it could be seen as a bug but it should be resolved in a subsequent PR.

Comment (olszomal) [reply crypto/pkcs7/pk7_doit.c:68]: Agree with @t8m

**Comment by vdukhovni** [comment]
@olszomal I've dismissed the outdated reviews requesting changes, so this should be ready to merge shortly. I would however be happy see you comment on the ASN1_STRING issue around the NDEF flag.

**Comment by mattcaswell** [comment]
Pushed to master. Thank you for your contribution.

**Pull request closed without merging**
